cmd/checkheader/main.go

```diff
@@ -0,0 +1,238 @@
+		if addDirective(&b, l) {
+			continue
+		}
+		if convertTypedef(&b, l) {
+			continue
+		}
+		if checkMacro(&b, l, i) {
+			i++
+		}
+	}
+	if err := sc.Err(); err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+func addDirective(w io.Writer, l string) bool {
+	if strings.HasPrefix(l, "// #include ") ||
+		strings.HasPrefix(l, "// #if ") ||
+		strings.HasPrefix(l, "// #endif") {
+		fmt.Fprintln(w, l[len("// "):])
+		return true
+	}
+	return false
+}
+
+// convertTypedef converts a typedef contained in the line l
+// into a #define pointing to the corresponding OpenSSL type.
+// Only void* typedefs starting with GO are converted.
+// If l does not contain a typedef it does nothing and returns false.
+func convertTypedef(w io.Writer, l string) bool {
+	if !strings.HasPrefix(l, "typedef void* GO_") {
+		return false
+	}
+	// Replace custom opaque pointer typedef with the proper OpenSSL type
+	// so gcc does not complain about pointer mismatch.
+	i1 := strings.Index(l, "GO_")
+	i2 := strings.Index(l, "_PTR")
+	if i2 < 0 {
+		log.Println("unexpected line in typedef: " + l)
+		return false
+	}
+	name := l[i1+3 : i2]
+	fmt.Fprintf(w, "#define GO_%s_PTR %s*\n", name, name)
+	return true
+}
+
+// checkMacro adds a static check which verifies that
+// the function definition macro contained in the line l
+// matches the corresponding OpenSSL function signature.
+// If l does not contain a function definition macro
+// it does nothing and returns false.
+func checkMacro(w io.Writer, l string, i int) bool {
+	var versionCond string
+	if strings.HasPrefix(l, "/*check:from=") {
+		i1 := strings.Index(l, "=")
+		i2 := strings.Index(l, "*/")
+		if i1 < 0 || i2 < 0 {
+			log.Println("unexpected 'check:from' condition: " + l)
+			return false
+		}
+		from := l[i1+1 : i2]
+		switch from {
+		case "1.1.0":
+			versionCond = "OPENSSL_VERSION_NUMBER >= 0x10100000L"
+		case "1.1.1":
+			versionCond = "OPENSSL_VERSION_NUMBER >= 0x10101000L"
+		case "3.0.0":
+			versionCond = "OPENSSL_VERSION_NUMBER >= 0x30000000L"
+		default:
+			log.Println("unexpected 'check:from' version" + l)
+			return false
+		}
+		l = l[i2+3:]
+	}
+	if !strings.HasPrefix(l, "DEFINEFUNC") {
+		return false
+	}
+	i1 := strings.IndexByte(l, '(')
+	// The first ")," match is always the end of the argument list parameter.
+	// We are not interested in the last parameter and parsing them would complicate the algorithm.
+	// Matching against ')' is not enough as it when the argument list parameter contains function pointers.
+	i2 := strings.Index(l, "),")
+	if i1 < 0 || i2 < 0 {
+		log.Println("unexpected line in function: " + l)
+		return false
+	}
+	subs := l[i1+1 : i2+1]
+	writeCheck := func(ret, name, args string) {
+		fmt.Fprintf(w, "__typeof__(%s(*)%s) __check_%d = %s;\n", ret, args, i, name)
+	}
+	writeDefineFunc := func(cond string) {
+		sp := strings.SplitN(subs, ",", 3)
+		if cond != "" {
+			fmt.Fprintf(w, "#if %s\n", cond)
+		}
+		writeCheck(sp[0], sp[1], sp[2])
```

Comment:
`strings.SplitN` returns an slice with N elements, regardless if the strings can be split in N substrings, so it won't go out of bounds.

I'll print a warning message in case any of substrings are empty.

Reply:
I tried it out earlier to make sure, it doesn't seem to with this example, at least:
```go
	s := strings.SplitN("a,a", ",", 50)
	fmt.Printf("%q %v\n", s, len(s))
	_ = s[40]
```
https://go.dev/play/p/LaFnIKUiLUM
```
["a" "a"] 2
panic: runtime error: index out of range [40] with length 2

goroutine 1 [running]:
main.main()
	/tmp/sandbox4130164030/prog.go:11 +0xda

Program exited.
```

Reply:
Wow, I was completely wrong! I'll add a check then just in case.